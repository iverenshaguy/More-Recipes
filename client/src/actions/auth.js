import instance from '../axios';
import { errorHandler } from '../utils';
import { uploading, uploadSuccess, uploadFailure } from './uploadImage';
import { auth as authAPI } from '../services/api/users';
import {
  AUTHENTICATED,
  UNAUTHENTICATED,
  AUTHENTICATION_ERROR,
  AUTHENTICATING,
  LOGIN_SUCCESS,
  LOGIN_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_ERROR,
  CLEAR_AUTH_ERROR,
  UPDATE_USER_IMAGE_SUCCESS,
  UPDATE_USER_IMAGE_FAILURE
} from './actionTypes';

const authenticating = () => ({
  type: AUTHENTICATING
});

const authenticationSuccess = data => ({
  type: AUTHENTICATED,
  payload: data
});

const authenticationFailure = error => ({
  type: AUTHENTICATION_ERROR,
  payload: error
});

const loginSuccess = data => ({
  type: LOGIN_SUCCESS,
  payload: data
});

const loginFailure = error => ({
  type: LOGIN_ERROR,
  payload: error
});

const signupSuccess = data => ({
  type: SIGNUP_SUCCESS,
  payload: data
});

const signupFailure = error => ({
  type: SIGNUP_ERROR,
  payload: error
});

const updateUserImageSuccess = user => ({
  type: UPDATE_USER_IMAGE_SUCCESS,
  payload: user
});

const updateUserImageFailure = error => ({
  type: UPDATE_USER_IMAGE_FAILURE,
  payload: error
});

const clearAuthError = () => ({
  type: CLEAR_AUTH_ERROR
});

const resetUser = () => ({
  type: UNAUTHENTICATED
});

const updateUserImage = (id, url) => async (dispatch) => {
  try {
    dispatch(uploading());
    const response = await instance.put(`/users/${id}`, { profilePic: url });
    dispatch(updateUserImageSuccess(response.data));
    dispatch(uploadSuccess());
  } catch (error) {
    const errorResponse = errorHandler(error);
    dispatch(updateUserImageFailure(errorResponse.response));
    dispatch(uploadFailure());
  }
};

const auth = type => user => async (dispatch) => {
  try {
    dispatch(authenticating());

    const response = await authAPI(type)(user);

    localStorage.setItem('jwtToken', response.data.token);

    // if type is login dispatch loginSuccess else dispatch signupSuccess
    const dispatchType = type === 'login' ? loginSuccess : signupSuccess;
    dispatch(dispatchType(response.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    // if type is login dispatch loginFailure else dispatch signupFailure
    const dispatchType = type === 'login' ? loginFailure : signupFailure;
    dispatch(dispatchType(errorResponse.response));
  }
};

const authenticateUser = () => async (dispatch) => {
  try {
    dispatch(authenticating());

    const res = await instance.get('/users/token');

    localStorage.setItem('jwtToken', res.data.token);

    dispatch(authenticationSuccess(res.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    localStorage.removeItem('jwtToken');

    dispatch(authenticationFailure(errorResponse.response));
  }
};

const logout = () => (dispatch) => {
  localStorage.removeItem('jwtToken');
  dispatch(resetUser());
};

export default {
  auth,
  loginSuccess,
  loginFailure,
  signupSuccess,
  signupFailure,
  authenticating,
  authenticateUser,
  authenticationSuccess,
  authenticationFailure,
  updateUserImage,
  updateUserImageSuccess,
  updateUserImageFailure,
  clearAuthError,
  resetUser,
  logout
};
