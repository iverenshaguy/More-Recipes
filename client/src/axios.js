import axios from 'axios';

// configure base url
const instance = axios.create({
  baseURL: '/api/v1',
});

// intercept requests, add authorization token
// and validate status function
instance.interceptors.request.use((config) => {
  const token = localStorage.getItem('jwtToken');
  config.headers.authorization = `Bearer ${token}`;
  config.validateStatus = status => status >= 200 && status < 300;
  return config;
});

// intercept response and reload page if request error
// is caused by an expired token
instance.interceptors.response.use(
  response => response,
  (error) => {
    const { response: { status, data } } = error;

    if (status === 403 &&
      data.error === 'User authorization token is expired') {
      window.location.reload();
    }

    return Promise.reject(error);
  }
);

export default instance;
