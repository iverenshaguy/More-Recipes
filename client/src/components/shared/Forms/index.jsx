import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form as BootstrapForm, Button } from 'reactstrap';
import { asyncValidate, syncValidate } from '../../../helpers/validations';
import AuthForm from './AuthForm';
import { NormalAlert } from '../Alert';
import { arrayToObject } from '../../../utils';
import { formHelpers } from '../../../helpers';

/**
 * @exports
 * @class Form
 * @extends Component
 * @returns {component} Form
 */
class Form extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
    type: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      title: PropTypes.string,
      btnText: PropTypes.string,
      extra: PropTypes.element
    }).isRequired
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof Form
   * @param {object} props - props
   * @returns {nothing} Return nothing
   */
  constructor(props) {
    super(props);
    const { type } = props;
    const { formFields } = formHelpers;
    const fields = formFields[type];

    this.state = {
      type,
      values: arrayToObject(fields, ''),
      touched: arrayToObject(fields, false),
      error: arrayToObject(fields, null),
      pristine: true,
      formValid: false,
      asyncValidating: false
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.validateField = this.validateField.bind(this);
    this.validateForm = this.validateForm.bind(this);
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  componentWillMount() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  clearFormErrors() {
    const { type } = this.state;
    const { clearFormError } = formHelpers;

    this.props.dispatch(clearFormError[type]);
  }


  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;

    this.setState({
      values: { ...this.state.values, [name]: value },
      touched: { ...this.state.touched, [name]: true },
      pristine: false,
    }, () => { this.validateField(name, value); });
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  handleFocus() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur(event) {
    const { target } = event;
    const { value, name } = target;
    const { type } = this.state;

    this.setState({
      touched: { ...this.state.touched, [name]: true },
    }, () => { this.validateField(name); });

    setTimeout(() => {
      if ((name === 'email' || name === 'username') && this.state.error[name] === null) {
        // only set asyncvalidating when sync validation is completed
        this.setState({ asyncValidating: true }, () => {
          asyncValidate(type)(name, value)
            .then(() => { this.setState({ asyncValidating: false }); })
            .catch((error) => {
              this.setState({
                error: { ...this.state.error, [name]: error[name] },
                asyncValidating: false
              });
            });
        });
      }
    }, 500);
  }

  /**
   * @memberof Form
   * @param {string} field
   * @param {string} values
   * @returns {nothing} Returns nothing
   */
  validateField(field) {
    const { type } = this.state;
    const error = syncValidate(type)(field, this.state.values);

    if (error) {
      this.setState({
        error: {
          ...this.state.error,
          [field]: error
        }
      }, this.validateForm);
    } else {
      this.setState({
        error: {
          ...this.state.error,
          [field]: null
        }
      }, this.validateForm);
    }
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  validateForm() {
    const formErrorArrayLength =
      Object.values(this.state.error).filter(value => value !== null).length;
    const formTouchedArrayLength =
      Object.values(this.state.touched).filter(value => value === true).length;
    const formFieldCount = Object.keys(this.state.values).length;

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength && formTouchedArrayLength === formFieldCount) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit(event) {
    event.preventDefault();
    const { type } = this.state;
    const { formSubmitMapper } = formHelpers;

    this.props.dispatch(formSubmitMapper[type](this.state.values));
  }

  /**
   * @memberof Form
   * @returns {component} Form
   */
  render() {
    const {
      values, touched, error, pristine, formValid, asyncValidating
    } = this.state;

    const {
      submitting, submitError, type, meta
    } = this.props;

    const { title, btnText, extra } = meta;

    const formState = {
      values,
      touched,
      error,
      pristine,
      formValid,
      asyncValidating
    };

    const handlers = {
      handleChange: this.handleChange,
      handleBlur: this.handleBlur,
      handleFocus: this.handleFocus,
      handleSubmit: this.handleSubmit
    };

    return (
      <div>
        <h4 className="text-center">{title}</h4>
        <hr />
        {type !== 'login' &&
          <p className="text-muted mx-auto text-center">
            Fields marked
            <span className="text-danger">*</span> are important
          </p>}
        <BootstrapForm className="mt-4 mb-3 px-5" onSubmit={handlers.handleSubmit}>
          {submitError && (
            <NormalAlert color="danger">
              <p className="text-center mb-0">{submitError}</p>
            </NormalAlert>
          )}
          {(type === 'signup' || type === 'login') && <AuthForm type={type} state={formState} handlers={handlers} />}
          <Button className="btn-block mt-0" disabled={!formValid || pristine || submitting}>
            {btnText}
          </Button>
        </BootstrapForm>
        {extra && extra}
      </div>);
  }
}


export { Form as FormComponent };

export default connect()(Form);
