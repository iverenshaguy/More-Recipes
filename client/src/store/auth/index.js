import reducer from './reducers';
import authOperations from './operations';
import authSelectors from './selectors';
import authTypes from './types';

export { authOperations, authSelectors, authTypes };

export default reducer;
