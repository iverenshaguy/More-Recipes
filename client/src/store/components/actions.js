import { TOGGLE_MODAL } from './types';

const toggleModal = modal => ({
  type: TOGGLE_MODAL,
  payload: modal || ''
});

export default {
  toggleModal
};
