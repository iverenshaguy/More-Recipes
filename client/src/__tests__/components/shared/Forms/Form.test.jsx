import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import axios from 'axios';
import moxios from 'moxios';
import { FormComponent } from '../../../../components/shared/Forms';
import { clearAuthError } from '../../../../actions/auth';

const dispatchMock = jest.fn();

const state = {
  type: 'login',
  values: {
    email: '',
    password: ''
  },
  touched: { email: false, password: false },
  error: { email: null, password: null },
  pristine: true,
  formValid: false,
  asyncValidating: false
};

const setup = () => {
  const props = {
    submitting: false,
    submitError: null,
    type: 'login',
    meta: {
      title: 'Sign In to Your Account',
      btnText: 'SIGN IN',
      extra: <p>something</p>
    }
  };

  const mountRoot = mount( //eslint-disable-line
    <MemoryRouter>
      <FormComponent {...props} dispatch={dispatchMock} />
    </MemoryRouter>);

  return { props, mountRoot };
};

const signupMeta = {
  title: 'Register for a New Account',
  btnText: 'SIGN UP',
  extra: <p>something</p>
};

const url = '/api/v1';

describe('Form', () => {
  beforeEach(() => {
    moxios.install(axios);
  });

  afterEach(() => {
    moxios.uninstall(axios);
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { props } = setup();
    const shallowComponent = shallow(<FormComponent {...props} dispatch={dispatchMock} />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
    expect(dispatchMock).toHaveBeenCalled();
    expect(dispatchMock.mock.calls[0]).toEqual([clearAuthError()]);
  });

  it('renders signup form when signup is passed', () => {
    const { props } = setup();
    const shallowComponent = shallow(<FormComponent
      {...props}
      type="signup"
      meta={signupMeta}
      dispatch={dispatchMock}
    />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
  });


  it('disables submit button when form is clean', () => {
    const { props } = setup();
    const shallowComponent = shallow(<FormComponent {...props} dispatch={dispatchMock} />);

    expect(shallowComponent.find('Button[disabled=true]')).toBeTruthy();
  });

  it('disables submit button when submitting form', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      submitting: true
    };
    const shallowComponent = shallow(<FormComponent {...newProps} dispatch={dispatchMock} />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
    expect(shallowComponent.find('Button[disabled=true]')).toBeTruthy();
  });

  it('shows error alert and disables submit button when there\'s a submit error', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      submitError: 'Username/Password do not match'
    };

    const shallowComponent = shallow(<FormComponent {...newProps} dispatch={dispatchMock} />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
    expect(shallowComponent.find('NormalAlert')).toBeTruthy();
    expect(shallowComponent.find('Button[disabled=true]')).toBeTruthy();
  });


  describe('test for right input', () => {
    it('calls handleChange and handleBlur on input change and blur for email field', (done) => {
      const { mountRoot } = setup();
      const wrapper = mountRoot.find(FormComponent);

      const changeState = {
        ...state,
        values: { ...state.values, email: 'iverenshaguy@gmail.com' },
        touched: { ...state.touched, email: true },
        error: { email: null, password: null },
        pristine: false
      };

      const blurState = {
        ...changeState,
        asyncValidating: true
      };

      const event = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };

      expect(toJson(wrapper)).toMatchSnapshot();

      wrapper.find('input[name="email"]').simulate('focus');
      expect(dispatchMock).toHaveBeenCalledWith(clearAuthError());

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="email"]').simulate('blur', event);

      setTimeout(() => {
        try {
          expect(wrapper.instance().state).toEqual(blurState);
          done();
        } catch (e) {
          done.fail(e);
        }
      }, 600);
    });

    it('doesn\'t async validate password field', (done) => {
      const { mountRoot } = setup();
      const wrapper = mountRoot.find(FormComponent);


      const changeState = {
        ...state,
        values: { ...state.values, password: 'iverenshaguy' },
        touched: { ...state.touched, password: true },
        error: { email: null, password: null },
        pristine: false
      };

      const event = { target: { name: 'password', value: 'iverenshaguy' } };

      wrapper.find('input[name="password"]').simulate('focus');
      expect(dispatchMock).toHaveBeenCalledWith(clearAuthError());

      wrapper.find('input[name="password"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="password"]').simulate('blur', event);

      setTimeout(() => {
        try {
          expect(wrapper.instance().state).toEqual(changeState);
          done();
        } catch (e) {
          done.fail(e);
        }
      }, 500);
    });

    it('sets asyncValidating to false if asyncValidate test passes', (done) => {
      const { mountRoot } = setup();
      const wrapper = mountRoot.find(FormComponent);


      moxios.stubRequest(`${url}/users/signin`, {
        status: 422,
        response: {
          errors: {
            password: { msg: 'Password must be specified' }
          }
        },
      }, 5);

      const changeState = {
        ...state,
        values: { ...state.values, email: 'iverenshaguy@gmail.com' },
        touched: { ...state.touched, email: true },
        error: { email: null, password: null },
        pristine: false
      };

      const blurState = {
        ...changeState,
        asyncValidating: false
      };

      const event = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };

      expect(toJson(wrapper)).toMatchSnapshot();

      wrapper.find('input[name="email"]').simulate('focus');
      expect(dispatchMock).toHaveBeenCalledWith(clearAuthError());

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="email"]').simulate('blur', event);

      setTimeout(() => {
        try {
          expect(wrapper.instance().state).toEqual(blurState);
          done();
        } catch (e) {
          done.fail(e);
        }
      }, 2000);
    });

    it('submits valid form', () => {
      const { mountRoot } = setup();
      const wrapper = mountRoot.find(FormComponent);

      const newState = {
        ...state,
        values: { email: 'iverenshaguy@gmail.com', password: 'iverenshaguy' },
        touched: { email: true, password: true },
        error: { email: null, password: null },
        pristine: false,
        formValid: true
      };

      const emailEvent = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };
      const passwordEvent = { target: { name: 'password', value: 'iverenshaguy' } };

      wrapper.find('input[name="email"]').simulate('focus');
      wrapper.find('input[name="email"]').simulate('change', emailEvent);
      wrapper.find('input[name="password"]').simulate('focus');
      wrapper.find('input[name="password"]').simulate('change', passwordEvent);

      expect(wrapper.instance().state).toEqual(newState);

      wrapper.find('form').simulate('submit', { preventDefault() { } });
      expect(dispatchMock).toHaveBeenCalled();

      mountRoot.unmount();
    });
  });

  describe('test for wrong input', () => {
    it('async validates field and form on input change and blur', (done) => {
      const { mountRoot } = setup();
      const wrapper = mountRoot.find(FormComponent);


      const changeState = {
        ...state,
        values: { ...state.values, email: 'emilysanders@gmail.com' },
        touched: { ...state.touched, email: true },
        error: { email: null, password: null },
        pristine: false
      };

      const blurState = {
        ...changeState,
        asyncValidating: false,
        error: { email: 'This email is not registered, please signup instead', password: null },
      };

      const event = { target: { name: 'email', value: 'emilysanders@gmail.com' } };

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="email"]').simulate('blur', event);

      moxios.stubRequest(`${url}/users/signin`, {
        status: 422,
        response: {
          errors: {
            email: { msg: 'This email is not registered, please signup instead' },
            password: { msg: 'Password must be specified' }
          }
        },
      }, 5);

      setTimeout(() => {
        try {
          expect(wrapper.instance().state).toEqual(blurState);
          done();
        } catch (e) {
          done.fail(e);
        }
      }, 800);
    });

    it('sync validates field and form on input change and blur', () => {
      const { mountRoot } = setup();
      const wrapper = mountRoot.find(FormComponent);


      const changeState = {
        ...state,
        values: { ...state.values, email: 'emilysanders' },
        touched: { ...state.touched, email: true },
        error: { email: 'Invalid email address!', password: null },
        pristine: false
      };

      const blurState = {
        ...changeState,
        asyncValidating: false
      };

      const event = { target: { name: 'email', value: 'emilysanders' } };

      wrapper.find('input[name="email"]').simulate('focus');
      wrapper.find('input[name="email"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(({
        ...changeState, error: { email: 'Required!', password: null }, pristine: true, values: { ...state.values }
      }));

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="email"]').simulate('blur', event);
      expect(wrapper.instance().state).toEqual(blurState);

      wrapper.find('input[name="password"]').simulate('focus');
      wrapper.find('input[name="password"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(({
        ...changeState,
        error: { ...changeState.error, password: 'Required!' },
        touched: { email: true, password: true },
        values: { email: 'emilysanders', password: '' }
      }));

      mountRoot.unmount();
    });
  });
});
