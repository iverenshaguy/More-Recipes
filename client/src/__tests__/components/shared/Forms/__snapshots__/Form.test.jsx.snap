// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form disables submit button when submitting form 1`] = `
<div>
  <h4
    className="text-center"
  >
    Sign In to Your Account
  </h4>
  <hr />
  <Form
    className="mt-4 mb-3 px-5"
    onSubmit={[Function]}
    tag="form"
  >
    <AuthForm
      handlers={
        Object {
          "handleBlur": [Function],
          "handleChange": [Function],
          "handleFocus": [Function],
          "handleSubmit": [Function],
        }
      }
      state={
        Object {
          "asyncValidating": false,
          "error": Object {
            "email": null,
            "password": null,
          },
          "formValid": false,
          "pristine": true,
          "touched": Object {
            "email": false,
            "password": false,
          },
          "values": Object {
            "email": "",
            "password": "",
          },
        }
      }
      type="login"
    />
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN IN
    </Button>
  </Form>
  <p>
    something
  </p>
</div>
`;

exports[`Form renders correctly 1`] = `
<div>
  <h4
    className="text-center"
  >
    Sign In to Your Account
  </h4>
  <hr />
  <Form
    className="mt-4 mb-3 px-5"
    onSubmit={[Function]}
    tag="form"
  >
    <AuthForm
      handlers={
        Object {
          "handleBlur": [Function],
          "handleChange": [Function],
          "handleFocus": [Function],
          "handleSubmit": [Function],
        }
      }
      state={
        Object {
          "asyncValidating": false,
          "error": Object {
            "email": null,
            "password": null,
          },
          "formValid": false,
          "pristine": true,
          "touched": Object {
            "email": false,
            "password": false,
          },
          "values": Object {
            "email": "",
            "password": "",
          },
        }
      }
      type="login"
    />
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN IN
    </Button>
  </Form>
  <p>
    something
  </p>
</div>
`;

exports[`Form renders signup form when signup is passed 1`] = `
<div>
  <h4
    className="text-center"
  >
    Register for a New Account
  </h4>
  <hr />
  <p
    className="text-muted mx-auto text-center"
  >
    Fields marked
    <span
      className="text-danger"
    >
      *
    </span>
     are important
  </p>
  <Form
    className="mt-4 mb-3 px-5"
    onSubmit={[Function]}
    tag="form"
  >
    <AuthForm
      handlers={
        Object {
          "handleBlur": [Function],
          "handleChange": [Function],
          "handleFocus": [Function],
          "handleSubmit": [Function],
        }
      }
      state={
        Object {
          "asyncValidating": false,
          "error": Object {
            "aboutMe": null,
            "email": null,
            "firstname": null,
            "lastname": null,
            "occupation": null,
            "password": null,
            "passwordConfirm": null,
            "username": null,
          },
          "formValid": false,
          "pristine": true,
          "touched": Object {
            "aboutMe": false,
            "email": false,
            "firstname": false,
            "lastname": false,
            "occupation": false,
            "password": false,
            "passwordConfirm": false,
            "username": false,
          },
          "values": Object {
            "aboutMe": "",
            "email": "",
            "firstname": "",
            "lastname": "",
            "occupation": "",
            "password": "",
            "passwordConfirm": "",
            "username": "",
          },
        }
      }
      type="signup"
    />
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN UP
    </Button>
  </Form>
  <p>
    something
  </p>
</div>
`;

exports[`Form shows error alert and disables submit button when there's a submit error 1`] = `
<div>
  <h4
    className="text-center"
  >
    Sign In to Your Account
  </h4>
  <hr />
  <Form
    className="mt-4 mb-3 px-5"
    onSubmit={[Function]}
    tag="form"
  >
    <NormalAlert
      color="danger"
    >
      <p
        className="text-center mb-0"
      >
        Username/Password do not match
      </p>
    </NormalAlert>
    <AuthForm
      handlers={
        Object {
          "handleBlur": [Function],
          "handleChange": [Function],
          "handleFocus": [Function],
          "handleSubmit": [Function],
        }
      }
      state={
        Object {
          "asyncValidating": false,
          "error": Object {
            "email": null,
            "password": null,
          },
          "formValid": false,
          "pristine": true,
          "touched": Object {
            "email": false,
            "password": false,
          },
          "values": Object {
            "email": "",
            "password": "",
          },
        }
      }
      type="login"
    />
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN IN
    </Button>
  </Form>
  <p>
    something
  </p>
</div>
`;

exports[`Form test for right input calls handleChange and handleBlur on input change and blur for email field 1`] = `
<Form
  dispatch={[Function]}
  meta={
    Object {
      "btnText": "SIGN IN",
      "extra": <p>
        something
      </p>,
      "title": "Sign In to Your Account",
    }
  }
  submitError={null}
  submitting={false}
  type="login"
>
  <div>
    <h4
      className="text-center"
    >
      Sign In to Your Account
    </h4>
    <hr />
    <Form
      className="mt-4 mb-3 px-5"
      onSubmit={[Function]}
      tag="form"
    >
      <form
        className="mt-4 mb-3 px-5"
        onSubmit={[Function]}
      >
        <AuthForm
          handlers={
            Object {
              "handleBlur": [Function],
              "handleChange": [Function],
              "handleFocus": [Function],
              "handleSubmit": [Function],
            }
          }
          state={
            Object {
              "asyncValidating": false,
              "error": Object {
                "email": null,
                "password": null,
              },
              "formValid": false,
              "pristine": true,
              "touched": Object {
                "email": false,
                "password": false,
              },
              "values": Object {
                "email": "",
                "password": "",
              },
            }
          }
          type="login"
        >
          <RenderInput
            handleBlur={[Function]}
            handleChange={[Function]}
            handleFocus={[Function]}
            id="email"
            label="Email Address"
            meta={
              Object {
                "asyncValidating": false,
                "error": null,
                "touched": false,
              }
            }
            name="email"
            placeholder="janesmith@me.com"
            required={false}
            type="email"
            value=""
          >
            <FormGroup
              tag="div"
            >
              <div
                className="form-group"
              >
                <Label
                  className="col-form-label"
                  for="email"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <label
                    className="col-form-label  form-control-label"
                    htmlFor="email"
                  >
                    Email Address
                  </label>
                </Label>
                <Input
                  className=""
                  id="email"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="janesmith@me.com"
                  tag="p"
                  type="email"
                  value=""
                >
                  <input
                    className="form-control"
                    id="email"
                    name="email"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="janesmith@me.com"
                    type="email"
                    value=""
                  />
                </Input>
              </div>
            </FormGroup>
          </RenderInput>
          <RenderInput
            className="mb-4"
            handleBlur={[Function]}
            handleChange={[Function]}
            handleFocus={[Function]}
            id="password"
            label="Password"
            meta={
              Object {
                "error": null,
                "touched": false,
              }
            }
            name="password"
            placeholder="••••••••••"
            required={false}
            type="password"
            value=""
          >
            <FormGroup
              tag="div"
            >
              <div
                className="form-group"
              >
                <Label
                  className="col-form-label"
                  for="password"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <label
                    className="col-form-label  form-control-label"
                    htmlFor="password"
                  >
                    Password
                  </label>
                </Label>
                <Input
                  className=""
                  id="password"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="••••••••••"
                  tag="p"
                  type="password"
                  value=""
                >
                  <input
                    className="form-control"
                    id="password"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="••••••••••"
                    type="password"
                    value=""
                  />
                </Input>
              </div>
            </FormGroup>
          </RenderInput>
        </AuthForm>
        <Button
          className="btn-block mt-0"
          color="secondary"
          disabled={true}
          tag="button"
        >
          <button
            className="btn-block mt-0 btn btn-secondary disabled"
            disabled={true}
            onClick={[Function]}
          >
            SIGN IN
          </button>
        </Button>
      </form>
    </Form>
    <p>
      something
    </p>
  </div>
</Form>
`;

exports[`Form test for right input sets asyncValidating to false if asyncValidate test passes 1`] = `
<Form
  dispatch={[Function]}
  meta={
    Object {
      "btnText": "SIGN IN",
      "extra": <p>
        something
      </p>,
      "title": "Sign In to Your Account",
    }
  }
  submitError={null}
  submitting={false}
  type="login"
>
  <div>
    <h4
      className="text-center"
    >
      Sign In to Your Account
    </h4>
    <hr />
    <Form
      className="mt-4 mb-3 px-5"
      onSubmit={[Function]}
      tag="form"
    >
      <form
        className="mt-4 mb-3 px-5"
        onSubmit={[Function]}
      >
        <AuthForm
          handlers={
            Object {
              "handleBlur": [Function],
              "handleChange": [Function],
              "handleFocus": [Function],
              "handleSubmit": [Function],
            }
          }
          state={
            Object {
              "asyncValidating": false,
              "error": Object {
                "email": null,
                "password": null,
              },
              "formValid": false,
              "pristine": true,
              "touched": Object {
                "email": false,
                "password": false,
              },
              "values": Object {
                "email": "",
                "password": "",
              },
            }
          }
          type="login"
        >
          <RenderInput
            handleBlur={[Function]}
            handleChange={[Function]}
            handleFocus={[Function]}
            id="email"
            label="Email Address"
            meta={
              Object {
                "asyncValidating": false,
                "error": null,
                "touched": false,
              }
            }
            name="email"
            placeholder="janesmith@me.com"
            required={false}
            type="email"
            value=""
          >
            <FormGroup
              tag="div"
            >
              <div
                className="form-group"
              >
                <Label
                  className="col-form-label"
                  for="email"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <label
                    className="col-form-label  form-control-label"
                    htmlFor="email"
                  >
                    Email Address
                  </label>
                </Label>
                <Input
                  className=""
                  id="email"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="janesmith@me.com"
                  tag="p"
                  type="email"
                  value=""
                >
                  <input
                    className="form-control"
                    id="email"
                    name="email"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="janesmith@me.com"
                    type="email"
                    value=""
                  />
                </Input>
              </div>
            </FormGroup>
          </RenderInput>
          <RenderInput
            className="mb-4"
            handleBlur={[Function]}
            handleChange={[Function]}
            handleFocus={[Function]}
            id="password"
            label="Password"
            meta={
              Object {
                "error": null,
                "touched": false,
              }
            }
            name="password"
            placeholder="••••••••••"
            required={false}
            type="password"
            value=""
          >
            <FormGroup
              tag="div"
            >
              <div
                className="form-group"
              >
                <Label
                  className="col-form-label"
                  for="password"
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <label
                    className="col-form-label  form-control-label"
                    htmlFor="password"
                  >
                    Password
                  </label>
                </Label>
                <Input
                  className=""
                  id="password"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="••••••••••"
                  tag="p"
                  type="password"
                  value=""
                >
                  <input
                    className="form-control"
                    id="password"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="••••••••••"
                    type="password"
                    value=""
                  />
                </Input>
              </div>
            </FormGroup>
          </RenderInput>
        </AuthForm>
        <Button
          className="btn-block mt-0"
          color="secondary"
          disabled={true}
          tag="button"
        >
          <button
            className="btn-block mt-0 btn btn-secondary disabled"
            disabled={true}
            onClick={[Function]}
          >
            SIGN IN
          </button>
        </Button>
      </form>
    </Form>
    <p>
      something
    </p>
  </div>
</Form>
`;
